cps/src/main/java/org/jenkinsci/plugins/workflow/cps/steps/ParallelStep.java

```diff
@@ -76,47 +107,70 @@ Callback callbackFor(String name) {
             @Override
             public void onSuccess(StepContext context, Object result) {
                 handler.outcomes.put(name, new Outcome(result, null));
-                checkAllDone();
+                checkAllDone(false);
             }
 
             @Override
             public void onFailure(StepContext context, Throwable t) {
                 handler.outcomes.put(name, new Outcome(null, t));
-                checkAllDone();
+                if (handler.originalFailure == null) {
+                    handler.originalFailure = new SimpleEntry<String, Throwable>(name, t);
+                }
+                checkAllDone(true);
             }
 
-            private void checkAllDone() {
+            private void checkAllDone(boolean stepFailed) {
                 Map<String,Object> success = new HashMap<String, Object>();
-                Entry<String,Outcome> failure = null;
                 for (Entry<String,Outcome> e : handler.outcomes.entrySet()) {
                     Outcome o = e.getValue();
 
-                    if (o==null)
-                        return; // some of the results are not yet ready
+                    if (o==null) {
+                        // some of the results are not yet ready
+                        if (stepFailed && handler.failFast && ! handler.isStopSent()) {
+                            handler.stopSent();
+                            try {
+                                handler.stepExecution.stop(new FailFastException());
```

Comment:
I'd also put the exception message and chain the original failure, since the exception can be visible.

It's bit like `InterruptedException`, that you'd often want to know who interrupted it.